torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -1028,7 +1008,12 @@ XLAGraphExecutor::ScheduleSyncTensorsGraph(
     const std::vector<torch::lazy::BackendDataPtr>& tensor_data_vec) {
   auto tensors_data =
       SetTensorData(tensors, coll->config, coll->indices, tensor_data_vec);
-  auto sharding_specs = CollectShardingSpecs(tensors, coll->indices);
+  std::vector<XLATensor::ShardingSpecPtr> sharding_specs(coll->indices.size(),
+                                                         nullptr);
+  // Extract sharding specs for the results and prepares the data placeholders.
+  ShardingUtil::PrepareOutputShardingPropagation(
+      tensors, coll->indices, cached_computation->computation, &tensors_data,
+      &sharding_specs);
```

Comment:
shouldn't we check whether graph is sharded before calling these functions?

Reply:
This call is a no-op if the graph is not sharded, as this is used in the SyncTensors code path.

Reply:
right, I would still prefer put this under a if condition, because
1. it is easy to read, I don't need to go into `PrepareOutputShardingPropagation` and figured that it is a no-op if it is not sharded
2. we don't need to allocate `sharding_specs` for non-SPMD
3. if one day we update `PrepareOutputShardingPropagation` and forgot about the caller always call(even for non-spmd) we might introduce weird issue.

Reply:
I see, good point. I think a better place would be to initialize and call `PrepareOutputShardingPropagation` if (async->is_sharded...)` just before the execution. Let me make some changes.

Reply:
make sure do that before the `barrier` call in https://github.com/pytorch/xla/blob/master/torch_xla/csrc/xla_graph_executor.cpp#L746, otherwise this will be blocked by the execution previous execution and can not be overlapped (which will increase the overhead of tracing in general).

Reply:
yea, also sourcing the inputs to the function seems untidy.. going with your original suggestion:
```
// Extract sharding specs for the results and prepare the sharded data
  // placeholders if the computation is sharded.
  if (cached_computation->is_sharded) {
    ShardingUtil::PrepareOutputShardingPropagation(
        tensors, coll->indices, cached_computation->computation, &tensors_data,
        &sharding_specs);
  }
```
we still need to initialize the sharding_specs with null, in case of a unsharded compuation.